File: 3-longest-substring-without-repeating-characters/solution.java
/*
 * LeetCode #3 - Longest Substring Without Repeating Characters
 * Difficulty : Medium
 * Language   : java
 * Runtime    : 5 ms
 * Memory     : 45.76 MB
 * URL        : https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> lastSeen = new HashMap<>();
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);

            if (lastSeen.containsKey(ch) && lastSeen.get(ch) >= left) {
                left = lastSeen.get(ch) + 1;
            }

            lastSeen.put(ch, right);
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}

File: 3-longest-substring-without-repeating-characters/solution.py
# LeetCode #3 - Longest Substring Without Repeating Characters
# Difficulty : Medium
# Language   : python
# Runtime    : 23 ms
# Memory     : 12.63 MB
# URL        : https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution:
    def lengthOfLongestSubstring(self, s):
        last_seen = {}
        left = 0
        max_len = 0
        #hello
        for right, ch in enumerate(s):
            if ch in last_seen and last_seen[ch] >= left:
                left = last_seen[ch] + 1

            last_seen[ch] = right
            max_len = max(max_len, right - left + 1)

        return max_len